src/configurator/v1.ml

```diff
@@ -283,42 +308,76 @@ module C_define = struct
       | String of string
   end
 
-  let import t ?c_flags ?link_flags ~includes vars =
+  let extract_program includes vars =
+    let has_type t = List.exists vars ~f:(fun (_, t') -> t = t') in
     let buf = Buffer.create 1024 in
     let pr fmt = Printf.bprintf buf (fmt ^^ "\n") in
-    let includes = "stdio.h" :: includes in
     List.iter includes ~f:(pr "#include <%s>");
-    pr "";
-    pr "int main()";
-    pr "{";
-    List.iter vars ~f:(fun (name, (kind : Type.t)) ->
-      match kind with
-      | Switch ->
-        pr {|#if defined(%s)|} name;
-        pr {|  printf("%s=b:true\n");|} name;
-        pr {|#else|};
-        pr {|  printf("%s=b:false\n");|} name;
-        pr {|#endif|}
-      | Int ->
-        pr {|  printf("%s=i:%%d\n", %s);|} name name
+    if has_type Type.Int then (
+      pr {|
+#define D0(x) ('0'+(x/1         )%%10)
+#define D1(x) ('0'+(x/10        )%%10), D0(x)
+#define D2(x) ('0'+(x/100       )%%10), D1(x)
+#define D3(x) ('0'+(x/1000      )%%10), D2(x)
+#define D4(x) ('0'+(x/10000     )%%10), D3(x)
+#define D5(x) ('0'+(x/100000    )%%10), D4(x)
+#define D6(x) ('0'+(x/1000000   )%%10), D5(x)
+#define D7(x) ('0'+(x/10000000  )%%10), D6(x)
+#define D8(x) ('0'+(x/100000000 )%%10), D7(x)
+#define D9(x) ('0'+(x/1000000000)%%10), D8(x)
+|}
+    );
+    if has_type Type.String then (
+      pr {|
+
+#if __USE_MINGW_ANSI_STDIO && defined(__MINGW64__)
+#define REAL_ARCH_INTNAT_PRINTF_FORMAT \"ll\"
+#else
+#define REAL_ARCH_INTNAT_PRINTF_FORMAT ARCH_INTNAT_PRINTF_FORMAT
+#endif
+|}
+    );
+    List.iteri vars ~f:(fun i (name, t) ->
+      let i = i + 1 in
+      match t with
+      | Type.Int ->
+        pr {|
+const char s%i[] = {
+  'B', 'E', 'G', 'I', 'N', '-',
+  D9((%s)),
+  '-', 'E', 'N', 'D'
+};
+|} i name
       | String ->
-        pr {|  printf("%s=s:%%s\n", %s);|} name name);
-    pr "  return 0;";
-    pr "}";
-    let code = Buffer.contents buf in
-    match compile_c_prog t ?c_flags ?link_flags code with
-    | Error () -> die "failed to compile program"
-    | Ok exe ->
-      run_capture_exn t ~dir:(Filename.dirname exe) (command_line exe [])
-      |> String.split_lines
-      |> List.map ~f:(fun s ->
-        let var, data = String.lsplit2_exn s ~on:'=' in
-        (var,
-         match String.lsplit2_exn data ~on:':' with
-         | "b", s -> Value.Switch (bool_of_string s)
-         | "i", s -> Int    (int_of_string s)
-         | "s", s -> String s
-         | _ -> assert false))
+        pr {|const char *s%ii = "BEGIN-" %s "-END"|} i name;
+      | Switch ->
+        pr {|
+#ifdef %s
+const char *s%i = "BEGIN-true-END";
+#else
+const char *s%i = "BEGIN-false-END";
+#endif
+|} name i i
+    );
+    Buffer.contents buf
+
+  let extract_values obj_file vars =
+    let obj = Io.read_file obj_file in
```

Comment:
What is the motivation to read the whole file into memory?  Granted, it is small so it is no harm but I am wondering...

Reply:
Fair enough. I pushed a commit to fix this.